https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=d88b917325546cee24a544dc05cbfbef0febfe25

From d88b917325546cee24a544dc05cbfbef0febfe25 Mon Sep 17 00:00:00 2001
From: Richard Biener <rguenther@suse.de>
Date: Fri, 18 Apr 2025 14:52:41 +0200
Subject: [PATCH] tree-optimization/119858 - type mismatch with POINTER_PLUS

The recent PFA early-break vectorization fix left us with a POINTER_PLUS
and non-sizetype offset.

	PR tree-optimization/119858
	* tree-vect-loop.cc (vectorizable_live_operation): Convert
	pointer offset to sizetype.
---
 gcc/tree-vect-loop.cc | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/gcc/tree-vect-loop.cc b/gcc/tree-vect-loop.cc
index 958b829fa8d1..2d35fa128864 100644
--- a/gcc/tree-vect-loop.cc
+++ b/gcc/tree-vect-loop.cc
@@ -11657,8 +11657,12 @@ vectorizable_live_operation (vec_info *vinfo, stmt_vec_info stmt_info,
 				    break_lhs_phi);
 
 		  if (POINTER_TYPE_P (TREE_TYPE (new_tree)))
-		    tmp2 = gimple_build (&iv_stmts, POINTER_PLUS_EXPR,
-					 TREE_TYPE (new_tree), new_tree, tmp2);
+		    {
+		      tmp2 = gimple_convert (&iv_stmts, sizetype, tmp2);
+		      tmp2 = gimple_build (&iv_stmts, POINTER_PLUS_EXPR,
+					   TREE_TYPE (new_tree), new_tree,
+					   tmp2);
+		    }
 		  else
 		    {
 		      tmp2 = gimple_convert (&iv_stmts, TREE_TYPE (new_tree),
-- 
2.43.5
